File: book/petrol-pump/chapter3/section9/PresetAmountPump.java
package chapter3.section9;

import pump.*;
import chapter3.section2.LifeCycle;
import chapter3.section2.LifeCycle.End;
import chapter3.section4.AccumulatePulsesPump;
import chapter3.section5.Fill;
import chapter3.section5.ShowDollarsPump;
import chapter3.section6.NotifyPointOfSale;
import chapter3.section7.Keypad;
import sodium.*;
import java.util.Optional;

public class PresetAmountPump implements Pump {
    public Outputs create(Inputs inputs) {
        EventLoop<Fuel> eStart = new EventLoop<>();
        Fill fi = new Fill(inputs.eClearSale.map(u -> Unit.UNIT),
                           inputs.eFuelPulses, inputs.calibration,
                           inputs.price1, inputs.price2, inputs.price3,
                           eStart);
        NotifyPointOfSale np = new NotifyPointOfSale(
                new LifeCycle(inputs.eNozzle1,
                              inputs.eNozzle2,
                              inputs.eNozzle3),
                inputs.eClearSale,
                fi);
        eStart.loop(np.eStart);
        BehaviorLoop<Boolean> keypadActive = new BehaviorLoop<>();
        Keypad ke = new Keypad(inputs.eKeypad,
                               inputs.eClearSale,
                               keypadActive);
        Preset pr = new Preset(ke.value,
                               fi,
                               np.fuelFlowing,
                               np.fillActive.map(o -> o.isPresent()));
        keypadActive.loop(pr.keypadActive);
        return new Outputs()
            .setDelivery(pr.delivery)
            .setSaleCostLCD(fi.dollarsDelivered.map(
                    q -> Formatters.formatSaleCost(q)))
            .setSaleQuantityLCD(fi.litersDelivered.map(
                    q -> Formatters.formatSaleQuantity(q)))
            .setPriceLCD1(ShowDollarsPump.priceLCD(np.fillActive, fi.price,
                    Fuel.ONE, inputs))
            .setPriceLCD2(ShowDollarsPump.priceLCD(np.fillActive, fi.price,
                    Fuel.TWO, inputs))
            .setPriceLCD3(ShowDollarsPump.priceLCD(np.fillActive, fi.price,
                    Fuel.THREE, inputs))
            .setSaleComplete(np.eSaleComplete)
            .setPresetLCD(ke.value.map(v ->
                Formatters.formatSaleCost((double)v)))
            .setBeep(np.eBeep.merge(ke.eBeep));
    }
}

File: c#/Book/PetrolPump/PetrolPump/PetrolPump/Chapter3/Section10/PresetAmountPump.cs
using PetrolPump.Chapter3.Section3;
using PetrolPump.Chapter3.Section6;
using PetrolPump.Chapter3.Section7;
using PetrolPump.Chapter3.Section8;
using PetrolPump.Pump;

using Sodium;

namespace PetrolPump.Chapter3.Section10
{

  public class PresetAmountPump : IPump
  {
    public Outputs Create(Inputs inputs) 
    {
        EventLoop<Fuel> eStart = new EventLoop<Fuel>();
        Fill fi = new Fill(inputs.EClearSale.Map(u => Unit.UNIT),
                           inputs.EFuelPulses, inputs.Calibration,
                           inputs.Price1, inputs.Price2, inputs.Price3,
                           eStart);
        NotifyPointOfSale np = new NotifyPointOfSale(
                new LifeCycle(inputs.ENozzle1,
                              inputs.ENozzle2,
                              inputs.ENozzle3),
                inputs.EClearSale,
                fi);
        eStart.Loop(np.EStart);
        BehaviorLoop<bool> keypadActive = new BehaviorLoop<bool>();
        Keypad ke = new Keypad(inputs.EKeypad,
                               inputs.EClearSale,
                               keypadActive);
        Preset pr = new Preset(ke.Value,
                               fi,
                               np.FuelFlowing,
                               np.FillActive.Map(o => o.IsPresent));
        keypadActive.Loop(pr.KeypadActive);
        return new Outputs()
            .SetDelivery(pr.Delivery)
            .SetSaleCostLcd(fi.DollarsDelivered.Map(q => Formatters.FormatSaleCost(q)))
            .SetSaleQuantityLcd(fi.LitersDelivered.Map(q => Formatters.FormatSaleQuantity(q)))
            .SetPriceLcd1(ShowDollarsPump.PriceLCD(np.FillActive, fi.Price, Fuel.ONE, inputs))
            .SetPriceLcd2(ShowDollarsPump.PriceLCD(np.FillActive, fi.Price, Fuel.TWO, inputs))
            .SetPriceLcd3(ShowDollarsPump.PriceLCD(np.FillActive, fi.Price, Fuel.THREE, inputs))
            .SetSaleComplete(np.ESaleComplete)
            .SetPresetLcd(ke.Value.Map(v => Formatters.FormatSaleCost((double)v)))
            .SetBeep(np.EBeep.Merge(ke.EBeep));
    }
  }
}
